libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+      __libcpp_is_trivially_relocatable_v<_Func> && is_trivially_destructible_v<_Func> &&
+      __fits_in_buffer<_Func,
+                       sizeof(_TriviallyDestructibleObject::__data_),
+                       alignof(_TriviallyDestructibleObject::__data_)>;
+
+public:
+  template <class _Func>
+  static constexpr bool __fits_in_trivially_relocatable_buffer =
+      __fits_in_trivially_relocatable_buffer_impl<remove_cvref_t<_Func>>;
+
+  template <class _Func>
+  static constexpr bool __fits_in_trivially_destructible_buffer =
+      __fits_in_trivially_destructible_buffer_impl<remove_cvref_t<_Func>>;
+};
+
+static_assert(alignof(__move_only_function_storage) == alignof(void*));
+static_assert(sizeof(__move_only_function_storage) == __move_only_function_storage::__target_size);
+
+template <class...>
+class move_only_function;
+
+template <class...>
+struct __is_move_only_function : false_type {};
+
+template <class _ReturnT, class... _ArgTypes>
+struct __is_move_only_function<move_only_function<_ReturnT(_ArgTypes...)>> : true_type {};
+
+_LIBCPP_END_NAMESPACE_STD
+
+// This header is only partially guarded on purpose. This header is an implementation detail of move_only_function.h
+// and generates multiple versions of std::move_only_function
+#endif // _LIBCPP___FUNCTIONAL_MOVE_ONLY_FUNCTION_IMPL_H
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV&
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT false
+#endif
+
+#define _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#ifdef _LIBCPP_ABI_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI [[_Clang::__trivial_abi__]]
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#endif
+
+template <class _ReturnT, class... _ArgTypes>
+class _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI move_only_function<_ReturnT(
+    _ArgTypes...) _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT)> {
+  using _Store = __move_only_function_storage;
+
+  template <class _Functor>
+  struct __function_wrappers_impl {
+    __function_wrappers_impl() = delete;
+    __function_wrappers_impl(const __function_wrappers_impl&) = delete;
+    __function_wrappers_impl(__function_wrappers_impl&&)      = delete;
+
+    _LIBCPP_HIDE_FROM_ABI static void __destroy(void* __store) noexcept {
+      std::destroy_at(static_cast<_Functor*>(__store));
+    }
+
+    _LIBCPP_HIDE_FROM_ABI static _ReturnT
+    __call(void* __functor, _ArgTypes&&... __args) noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
```

Comment:
Bikeshed: Remove the `noexcept`-spec here; `noexcept` on internal implementation details can't possibly help, and could hurt (by causing an unexpected `terminate` if it's gotten wrong by accident).